doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1588,39 +1588,70 @@ Controlling the proof flow
 
 .. tacn:: specialize @one_term_with_bindings {? @as_ipat }
 
-   This tactic works on the :term:`head constant` of the :n:`@one_term` in
-   :n:`@one_term_with_bindings`.
-   The premises of this hypothesis (either universal quantifications or
-   :term:`non-dependent premises <non-dependent premise>`) are instantiated by
-   concrete terms from :n:`@one_term_with_bindings`.
-
-   .. todo PR:
-      1) what is "non-dependent implication" above?
-      2) In this little example, what exactly is (H x)? It looks almost like
-         a function but H doesn't look like a function.  Is this just something
-         wierd about specialize or have I missed some basic point?
-
-      Goal forall x: nat, (forall n, n >= 0) -> True.
-      intros.  (* H is forall n : nat, n >= 0 *)
-      specialize (H x). (* H is x >= 0 *)
-
-   The application to :n:`@one_term` can be partial. The
-   uninstantiated arguments are inferred by unification if possible or left
-   quantified in the hypothesis otherwise. With the :n:`as` clause, the local
-   hypothesis :n:`@ident` is left unchanged and instead, the modified hypothesis
-   is introduced as specified by the :token:`simple_intropattern`. The name :n:`@ident`
-   can also refer to a global lemma or hypothesis. In this case, for
-   compatibility reasons, the behavior of :tacn:`specialize` is close to that of
-   :tacn:`generalize`: the instantiated statement becomes an additional premise of
-   the goal. The ``as`` clause is especially useful in this case to immediately
-   introduce the instantiated statement as a local hypothesis.
+   If the :term:`head constant` of :n:`@one_term` has the type
+   `forall ...`, the tactic replaces one or more of the
+   quantified variables in the type with terms provided by
+   :n:`@one_term_with_bindings`, either in the form of a
+   :ref:`function application <function_application>` (which may be partial),
+   such as `(H 1)`, or with named binders, such as `H with (n:=1)`.
+
+   If the :term:`head constant` has a non-dependent product type such as
+   `A -> B -> C`, the tactic does :term:`forward reasoning`, eliminating one or
+   more of the premises.
+
+   Uninstantiated arguments are inferred by unification, if possible, or otherwise
+   left quantified in the resulting term.
```

Comment:
This is true, but there is a difference between `specialize (X ...)` and `specialize X with ...`.
- in the form `specialize (t1 ... tn)`,  the term `(t1 ... tn)` needs to be well formed and well typed. In particular you cannot omit the ith argument if you want to give the i+1th one. So the only form of partial application is by omitting the *last* arguments of `H`. This is the usual curryfied partial application.
- in the form `specialize t1 with (3:=...) (6:=...)... .` You can give arbitrary arguments to `H`, eliminating for instance the third and sixth ones, and leave the others quantified (actually a few others may be instantiated by side effect). This form is therefore more powerful in the sens that it gives more control on picking the argments you want to instantiate.
Sorry for poor english here. Hope this is clear.

Reply:
I think being well formed and well typed is almost always required, so I don't want to repeat that here (or 100 other places).  The reader has work a little to learn pervasive facts about Coq.

I changed one of the examples to use a numbered argument, whose description is a click away at `one_term_with_bindings`.

> Sorry for poor english here. Hope this is clear.

More than good enough for our informal conversation and quite clear.